projects/plugins/jetpack/extensions/shared/execution-lock.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * Map that relates a locked key with its unique lock value.
+ *
+ * @type {Map<string, number>}
+ */
+const lockMap = new Map();
+
+/**
+ * Acquires a lock for a given key.
+ *
+ * @param {string} key - Key for which we will acquire a lock.
+ * @returns {null|[string,number]} - After successfully acquiring a lock we return the lock, if the key is locked, null is returned.
+ */
+const acquire = key => {
+	if ( isLocked( key ) ) {
+		return null;
+	}
+
+	const lockValue = Math.random();
+	lockMap.set( key, lockValue );
+	return [ key, lockValue ];
+};
+
+/**
+ * Block the execution until the locked key is released without blocking the thread.
+ *
+ * @param {string} key - Key that we will wait for release.
+ * @param {number} timeOffset - Amount of time to wait between lock checks.
+ */
+const blockExecution = async ( key, timeOffset = 10 ) => {
+	if ( isLocked( key ) ) {
+		await new Promise( resolve => setTimeout( resolve, timeOffset ) );
```

Comment:
Tried an exponential backoff approach at first, but it ended up being considerably slower than necessary so I've reverted to a fixed dt.